### PR TITLE
Emit coverage mappings for all modules

### PR DESCRIPTION
**Pull request by keith**
Previously in WMO builds where IR was multithreaded only the
primary module would emit the coverage mapping leading to only the first
object file to have the __llvm_covmap section. This change emits
coverage for all modules so they are correctly reflected in the final
coverage report.

This fixes https://forums.swift.org/t/code-coverage-with-wmo/32561/

Fixes rdar://42562585
Fixes rdar://42562307

### CODE REVIEW

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -1105,8 +1105,11 @@ void IRGenerator::emitGlobalTopLevel(llvm::StringSet<> *linkerDirectives) {
     IGM->emitSILDifferentiabilityWitness(&dw);
   }
 
-  // Emit code coverage mapping data.
-  PrimaryIGM->emitCoverageMapping();
+  // Emit code coverage mapping data for all modules
+  for (auto Iter : *this) {
+    IRGenModule *IGM = Iter.second;
+    IGM->emitCoverageMapping();
+  }
```
Comment (keith) [inline lib/IRGen/GenDecl.cpp:1112]: I'm not sure how order-sensitive all this code is so I left this here in its own loop but I'm happy to merge this into another one if that is safe!

**Comment by keith** [comment]
Updated with a test

File: test/Profiler/coverage_num_threads.swift [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// RUN: %target-swift-frontend -profile-generate -profile-coverage-mapping -num-threads 0 -emit-ir %S/Inputs/coverage_num_threads1.swift | %FileCheck %s -check-prefix=SINGLE-SOURCE
+
+// SINGLE-SOURCE: __LLVM_COV,__llvm_covmap
```
Comment (vedantk) [inline test/Profiler/coverage_num_threads.swift:3]: These section names are different on Windows. How about something like: `FileCheck --implicit-check-not=llvm_coverage_mapping`, `SINGLE-SOURCE: llvm_coverage_mapping`?

File: test/Profiler/coverage_num_threads.swift [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+// SINGLE-SOURCE-NOT: __LLVM_COV,__llvm_covmap
+
+// RUN: %target-swift-frontend -profile-generate -profile-coverage-mapping -num-threads 0 -emit-ir %S/Inputs/coverage_num_threads1.swift %S/Inputs/coverage_num_threads2.swift | %FileCheck %s -check-prefix=SINGLE-OBJECT
+
+// SINGLE-OBJECT: __LLVM_COV,__llvm_covmap
+// SINGLE-OBJECT-NOT: __LLVM_COV,__llvm_covmap
+
+// RUN: %target-swift-frontend -profile-generate -profile-coverage-mapping -num-threads 12 -emit-ir %S/Inputs/coverage_num_threads1.swift %S/Inputs/coverage_num_threads2.swift | %FileCheck %s -check-prefix=MULTIPLE-OBJECTS
```
Comment (vedantk) [inline test/Profiler/coverage_num_threads.swift:11]: Interesting, I'm not sure how modules get split up for parallel compilation. Is -num-threads=12 necessary here, or can a smaller number work?

File: lib/IRGen/GenDecl.cpp (hunk 1 above)
Comment (vedantk) [reply lib/IRGen/GenDecl.cpp:1112]: This looks fine to me, the steps here seem fairly decoupled.

**Review by vedantk — COMMENTED** [review]
Thank you!

File: test/Profiler/coverage_num_threads.swift (hunk 3 above)
Comment (keith) [reply test/Profiler/coverage_num_threads.swift:11]: any multi-threaded number should work, changed to 2

File: test/Profiler/coverage_num_threads.swift (hunk 2 above)
Comment (keith) [reply test/Profiler/coverage_num_threads.swift]: Ah right windows, originally I was doing this but `llvm_coverage_mapping` actually appears multiple times in the successful case, updated the test to this but with `llvm_coverage_mapping =` which should be the same

**Review by vedantk — APPROVED** [review]
Lgtm, thanks again for this. You may encounter issues with swift-ci testing at the moment; our servers in Cupertino are overheating (perhaps due to the weather outside..)

**Comment by keith** [comment]
I don't have permissions or anything so do you mind landing this?

**Comment by vedantk** [comment]
@swift-ci smoke test and merge

**Comment by vedantk** [comment]
I can't force merge, but I've tried to kick off a CI job which should integrate this.

**Comment by keith** [comment]
totally, thanks!

File: test/Profiler/coverage_num_threads.swift [hunk 4]
```diff
@@ -0,0 +1,12 @@
+// RUN: %target-swift-frontend -profile-generate -profile-coverage-mapping -num-threads 0 -emit-ir %S/Inputs/coverage_num_threads1.swift | %FileCheck %s -check-prefix=SINGLE-SOURCE --implicit-check-not="llvm_coverage_mapping ="
+
+// SINGLE-SOURCE: llvm_coverage_mapping =
+
+// RUN: %target-swift-frontend -profile-generate -profile-coverage-mapping -num-threads 0 -emit-ir %S/Inputs/coverage_num_threads1.swift %S/Inputs/coverage_num_threads2.swift | %FileCheck %s -check-prefix=SINGLE-OBJECT --implicit-check-not="llvm_coverage_mapping ="
+
+// SINGLE-OBJECT: llvm_coverage_mapping =
+
+// RUN: %target-swift-frontend -profile-generate -profile-coverage-mapping -num-threads 2 -emit-ir %S/Inputs/coverage_num_threads1.swift %S/Inputs/coverage_num_threads2.swift | %FileCheck %s -check-prefix=MULTIPLE-OBJECTS --implicit-check-not="llvm_coverage_mapping ="
```
Comment (vedantk) [inline test/Profiler/coverage_num_threads.swift:9]: Is there some way other than -num-threads to control how files are split up into different modules (maybe @eeckstein  knows)?

The issue on Windows appears to be that only the IR for one of the modules is printed out. The first '@__llvm_coverage_mapping =' definition is matched, but the second isn't.

Comment (keith) [reply test/Profiler/coverage_num_threads.swift:9]: If you're able to quickly test this (I'm spinning up a windows env now) does changing the thread count to 1 change this behavior? I originally did 2 just to make it super clear it was going through the multi-threaded path but 1 also does, so if that affected this that test would still be valid.

Comment (vedantk) [reply test/Profiler/coverage_num_threads.swift:9]: I don't have access to a Windows environment right now. And I'm not sure I follow your comment here: why should swift only emit the IR for one module, given two source files and -num-threads=2?

Comment (keith) [reply test/Profiler/coverage_num_threads.swift:9]: Oh I'm definitely not saying it should, it sounds like a bug in `-emit-ir` only for windows to me, so I was just thinking maybe `1` would avoid that.

Comment (keith) [reply test/Profiler/coverage_num_threads.swift:9]: For historians this was re-merged here https://github.com/apple/swift/pull/32352

**Pull request merged**